src/js/background/sync.js

```diff
@@ -0,0 +1,461 @@
+      //if this fails, there is probably differing sync info.
+      console.error("Error getting info on CI", error);
+    }
+  }
+  try {
+  // update the local container with the sync data
+    await browser.contextualIdentities
+      .update(localCookieStoreID, identityInfo);
+    return;
+  } catch (error) {
+    // If this fails, sync info is off.
+    console.error("Error udpating CI", error);
+  }
+}
+
+async function ifNoMatch(syncIdentity){
+  // if no uuid match either, make new identity
+  console.log("create new ident: ", syncIdentity.name);
+  const newIdentity = 
+        await browser.contextualIdentities.create({
+          name: syncIdentity.name, 
+          color: syncIdentity.color, 
+          icon: syncIdentity.icon
+        });
+  await identityState.updateUUID(
+    newIdentity.cookieStoreId, 
+    syncIdentity.macAddonUUID
+  );
+  return;
+}
+/*
+ * Checks for site previously assigned. If it exists, and has the same
+ * container assignment, the assignment is kept. If it exists, but has
+ * a different assignment, the user is prompted (not yet implemented).
+ * If it does not exist, it is created.
+ */
+async function reconcileSiteAssignments() {
+  console.log("reconcileSiteAssignments");
+  const assignedSitesLocal = 
+    await assignManager.storageArea.getAssignedSites();
+  const assignedSitesFromSync = 
+    await sync.storageArea.getAssignedSites();
+  const deletedSiteList = 
+    await sync.storageArea.getDeletedSiteList();
+  for(const siteStoreKey of deletedSiteList) {
+    if (Object.prototype.hasOwnProperty.call(assignedSitesLocal,siteStoreKey)) {
+      assignManager
+        .storageArea
+        .remove(siteStoreKey.replace(/^siteContainerMap@@_/, "https://"));
+    }
+  }
+
+  for(const urlKey of Object.keys(assignedSitesFromSync)) {
+    const assignedSite = assignedSitesFromSync[urlKey];
+    try{
+      const syncUUID = assignedSite.identityMacAddonUUID;
+      console.log("syncUUID, ", syncUUID);
+      if (assignedSite.identityMacAddonUUID) {
+      // Sync is truth.
+      // Not even looking it up. Just overwrite
+        if (SYNC_DEBUG){ 
+          const isInStorage = await assignManager.storageArea.getByUrlKey(urlKey);
+          if (!isInStorage)
+            console.log("new assignment ", assignedSite);
+        }
+
+        await setAssignmentWithUUID(assignedSite, urlKey);
+        continue;
+      }
+    } catch (error) {
+      // this is probably old or incorrect site info in Sync
+      // skip and move on.
+    }
+  }
+}
+
+async function setAssignmentWithUUID (assignedSite, urlKey) {
+  const uuid = assignedSite.identityMacAddonUUID;
+  const cookieStoreId = await identityState.lookupCookieStoreId(uuid);
+  if (cookieStoreId) {
+    // eslint-disable-next-line require-atomic-updates
+    assignedSite.userContextId = cookieStoreId
+      .replace(/^firefox-container-/, "");
+    await assignManager.storageArea.set(
+      urlKey.replace(/^siteContainerMap@@_/, "https://"),
```

Comment:
This replacement to `https://` is done in numerous places, I still am not sure if it's correct however it should probably be moved into it's own function.

Reply:
The regular expressions could be stored as constants, but otherwise I don't see any real use for a separate function call here.

Reply:
My rationales:
- Use of a well worded function name can help explain what the code is doing
- Reduce the copy and paste issues
- Simplify the review process if said code ever needs to account for different keys etc
- The current `storageArea` architecture is supposed to be isolated to code using it, even within it's own component, how it gets stored should only be understood by the storageArea functions such that there is separation of concerns.

Other parts of this code certainly don't abide to the last principal.

Reply:
Halfway through I chanced `assignManager.getSiteStoreKey()` to take the url or the original siteStoreKey, so probably not needed anymore.